Add HLL lexer

gabrielhdt opened the pull request:
This pull requests adds a lexer for the _High Level Language_ (a.k.a. HLL), a language used in formal verification for railway systems. The specification is available [here](https://hal.science/hal-03294999v1) or [there](https://hal.science/hal-03356342v1/).

gabrielhdt committed:
Add lexer for the HLL

See https://hal.science/hal-03294999v1 for more information
on the High Level Language

gabrielhdt committed:
Fix tests and add test sample

gabrielhdt committed:
Implement rubocop suggestions

gabrielhdt committed:
Simplify and complete lexer

gabrielhdt committed:
Add String lexing

gabrielhdt committed:
Add type keywords

gabrielhdt committed:
Fix single line comment

gabrielhdt committed:
Add Outputs keyword and downcase a constraints kw

gabrielhdt committed:
Copy the list of keywords from the spec

jneen commented on lib/rouge/lexers/hll.rb:
```diff
@@ -0,0 +1,107 @@
+                  u2bin
+                  unsigned
+                  with
+                  X)
```
I'm having trouble finding this in the docs - are these keywords case insensitive, or are *only* the cases listed here (like `true`, `True`, `TRUE`) valid? If it's case-insensitive, it's probably better to upcase or downcase before checking inclusion in this set.

gabrielhdt committed:
Add dot in operators (for projections)

gabrielhdt committed:
More stressing visual sample for HLL

gabrielhdt commented on spec/visual/samples/hll:
```diff
@@ -0,0 +1,19 @@
+  b := ALL i:[0, NOF_PIGEONS-1] SOME j:[0, NOF_HOLES-1] (P[i,j]);
+
+proof obligations:
+  ~(a & b);
```
I created a new stress test that should cover most common syntax constructions. Don't hesitate to tell me if you prefer an even more complete coverage.

gabrielhdt committed:
Bigger HLL sample, tests many more HLL grammar cases

gabrielhdt committed:
Lex some more particular structures

- Lex built in functions
- Lex more operators (those containing !)
- Lex binary and hexadecimal literals
